drivers/flash/spi_nand.c

```diff
@@ -0,0 +1,1044 @@
+
+#include <errno.h>
+#include <zephyr/drivers/flash.h>
+#include <zephyr/drivers/spi.h>
+#include <zephyr/init.h>
+#include <string.h>
+#include <zephyr/logging/log.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include "spi_nand.h"
+#if IS_ENABLED(CONFIG_SPI_NAND_SOFTWARE_ECC)
+#include "bch.h"
+#define SPI_NAND_ECC_SIZE CONFIG_SPI_NAND_ECC_SIZE
+#endif
+
+LOG_MODULE_REGISTER(spi_nand, CONFIG_FLASH_LOG_LEVEL);
+
+/* Indicates that an access command includes bytes for the address.
+ * If not provided the opcode is not followed by address bytes.
+ */
+#define NAND_ACCESS_ADDRESSED BIT(0)
+
+/* Indicates that addressed access uses a 8-bit address regardless of
+ * spi_nand_data::flag_8bit_addr.
+ */
+#define NAND_ACCESS_8BIT_ADDR BIT(1)
+
+/* Indicates that addressed access uses a 16-bit address regardless of
+ * spi_nand_data::flag_16bit_addr.
+ */
+#define NAND_ACCESS_16BIT_ADDR BIT(2)
+
+/* Indicates that addressed access uses a 24-bit address regardless of
+ * spi_nand_data::flag_32bit_addr.
+ */
+#define NAND_ACCESS_24BIT_ADDR BIT(3)
+
+/* Indicates that addressed access uses a 32-bit address regardless of
+ * spi_nand_data::flag_32bit_addr.
+ */
+#define NAND_ACCESS_32BIT_ADDR BIT(4)
+
+/* Indicates that an access command is performing a write.  If not
+ * provided access is a read.
+ */
+#define NAND_ACCESS_WRITE BIT(5)
+
+#define NAND_ACCESS_DUMMY BIT(6)
+
+#define OP_TYPE_RD	0
+#define OP_TYPE_PGM	1
+#define FLASH_WRITE_SIZE DT_PROP(DT_INST(0, DT_DRV_COMPAT), page_size)
+#define FLASH_SIZE DT_PROP(DT_INST(0, DT_DRV_COMPAT), flash_size)
+
+
+/* Build-time data associated with the device. */
+struct spi_nand_config {
+	/* Devicetree SPI configuration */
+	struct spi_dt_spec spi;
+	uint8_t id[2];
+	uint32_t page_size;
+};
+
+/**
+ * struct spi_nand_data - Structure for defining the SPI NAND access
+ * @sem: The semaphore to access to the flash
+ */
+struct spi_nand_data {
+	struct k_sem sem;
+
+	/* Number of bytes per page */
+	uint16_t page_size;
+
+	/* Number of oob bytes per page */
+	uint8_t oob_size;
+
+	/* Number of pages per block */
+	uint16_t page_num;
+
+	/* Number of bytes per block */
+	uint32_t block_size;
+
+	/* Number of blocks per chip */
+	uint16_t block_num;
+
+	/* Size of flash, in bytes */
+	uint64_t flash_size;
+
+	uint8_t page_shift;
+
+	uint8_t block_shift;
+
+	bool continuous_read;
+
+	uint16_t prg_timeout;
+
+	uint8_t *page_buf;
+	struct nand_ecc_info {
+		uint8_t ecc_bits;
+		uint8_t ecc_bytes;
+		uint8_t ecc_steps;
+		uint8_t ecc_layout_pos;
+		uint32_t ecc_size;
+		uint8_t *ecc_calc;
+		uint8_t *ecc_code;
+	} ecc;
+
+#if IS_ENABLED(CONFIG_SPI_NAND_SOFTWARE_ECC)
+	struct nand_bch_control {
+		bch_t      *bch;
+		uint8_t			  *mask_ff;
+		uint8_t			  *input_data;
+	} nbc;
+#endif
+};
+
+int fmsb32(uint32_t bits)
+{
+	int n = 32, m = 16;
+	uint32_t x = 0xffffffff;
+
+	if (!bits) {
+		return -1;
+	}
+
+	do {
+		if (!(bits & (x << (32 - m)))) {
+			bits <<= m;
+			n -= m;
+		}
+		m >>= 1;
+	} while(m);
```

Comment:
No implicit boolean condition.

```suggestion
	} while (m > 0);
```

Reply:
Resolved